services/QuillLMS/app/services/diagnostic_insights/activity_session_skill_performance.rb

```diff
@@ -0,0 +1,116 @@
+# frozen_string_literal: true
+
+module DiagnosticInsights
+  class ActivitySessionSkillPerformance < ApplicationService
+    attr_reader :activity_session
+
+    def initialize(activity_session)
+      @activity_session = activity_session
+    end
+
+    def run
+      return unless concept_results.present?
+      return unless diagnostic? || recommendation?
+
+      skill_performance
+    end
+
+    private def skill_performance
+      # NOTE: Because we've never consolidated these, "Skills" for Diagnostic activities are derived from the "SkillGroup" model,
+      # while "Skills" for recommended activities are derived from the "Recommendation" model.
+      # There is no direct database mapping linking these together, just the intentional coincidence of names
+      questions_correct
```

Comment:
It's safer to move the 

`return unless diagnostic? || recommendation?` guard

into this method, instead of `run`. If the implementation of `run` changes in the future, `skill_performance` could error, since `questions_correct` can return nil without the 

`return unless diagnostic? || recommendation?`  guard